### ISSUE TITLE
Unexpected unhandled NatsError: BAD_AUTHENTICATION

### ISSUE THREAD

**Issue by ArmorDarks** [nats-io/nats.js#524]
* Client version: 2.8.0
* Node version: 16.14.2

Some servers, after loosing the connection to the NATS node receive following error:

```
NatsError: BAD_AUTHENTICATION
    at Function.errorForCode (/code/node_modules/nats/lib/nats-base-client/error.js:98:16)
    at ProtocolHandler.processInfo (/code/node_modules/nats/lib/nats-base-client/protocol.js:402:47)
    at ProtocolHandler.push (/code/node_modules/nats/lib/nats-base-client/protocol.js:435:22)
    at Parser.parse (/code/node_modules/nats/lib/nats-base-client/parser.js:460:45)
    at ProtocolHandler.<anonymous> (/code/node_modules/nats/lib/nats-base-client/protocol.js:189:45)
    at Generator.next (<anonymous>)
    at fulfilled (/code/node_modules/nats/lib/nats-base-client/protocol.js:5:58)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
```

`chainedError` property of the error says that:

```
TypeError: Cannot read property 'write' of undefined
    at NodeTransport.send (/code/node_modules/nats/lib/src/node_transport.js:327:21)
    at ProtocolHandler.processInfo (/code/node_modules/nats/lib/nats-base-client/protocol.js:398:32)
    at ProtocolHandler.push (/code/node_modules/nats/lib/nats-base-client/protocol.js:435:22)
    at Parser.parse (/code/node_modules/nats/lib/nats-base-client/parser.js:460:45)
    at ProtocolHandler.<anonymous> (/code/node_modules/nats/lib/nats-base-client/protocol.js:189:45)
    at Generator.next (<anonymous>)
    at fulfilled (/code/node_modules/nats/lib/nats-base-client/protocol.js:5:58)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
```

As seen from the chained error, it's actually not an auth error but some race condition inside of the NATS base client, which encounters an error [here](https://github.com/nats-io/nats.js/blob/main/src/node_transport.ts#L332).

Since that error [forces connection close](https://github.com/nats-io/nats.deno/blob/177c3da18319cbd0ec6066228e08f6709feb0511/nats-base-client/protocol.ts#L516), it wrongly closes connection for good and NATS will no longer try to reconnect despite there's actually no auth error.

This is an extremely rare error I saw only happening during reconnection on the server. One of the servers started to throw `NatsError: DISCONNECTED`, and after a few attempts, threw `NatsError: BAD_AUTHENTICATION` and closed the connection.

<img width="607" alt="image" src="https://user-images.githubusercontent.com/4460311/196996495-e1b30dce-991d-441e-a6c5-970b0ca6ecc1.png">

This is how statuses changed:

<img width="335" alt="image" src="https://user-images.githubusercontent.com/4460311/196997020-81ef3b16-9db1-4415-baa2-2eaba37d4caa.png">

The issue is probably related to https://github.com/nats-io/nats.js/issues/523 because it happened simultaneously, though only on 1 machine.

**Comment by aricart** [comment]
@ArmorDarks are these servers direct connected or are they behind some other load balancer or something to that effect?

Reconnects will terminate anytime there's more than one consecutive authentication error.

- The rejection of the initial PONG possibly has an unexpected race condition that will require additional handling.
- The thing that I am not clear on is why is the close handler being surfaced before the read side errors out. As you may have seen clients don't typically fail on write, only on read. But here we have a write happening after the close handler has nullified the socket, and that I had not seen.

One question, are you wrapping the client with some socket library like DataDog or something to that effect that would audit the data from the socket?

@ArmorDarks since you seem to be able to reproduce this, could I post a fix that you could test for me?

I also would be interested in seeing a run where it fails and debug is set to true. That may yield more goodness.

**Comment by aricart** [comment]
One more - what authentication is configured on this cluster, is this a NATS resolver?

**Comment by aricart** [comment]
So the current node transport, nullifies the socket, and then calls _closed(), which actually would know and log the error. However because of the unexpected close event, it NPEs and that mechanism never reports on that. I am flipping the behaviour - the client will log the error (because currently there's not much it can do with it)

@ArmorDarks `npm install nats@next` if you would like to try the fix, and see if you run into the issue.

**Comment by ArmorDarks** [comment]
@aricart 

> are these servers direct connected or are they behind some other load balancer or something to that effect?

They are behind the balancer, so there's only 1 server URL specified in the config

> The thing that I am not clear on is why is the close handler being surfaced before the read side errors out. As you may have seen clients don't typically fail on write, only on read. But here we have a write happening after the close handler has nullified the socket, and that I had not seen.

I'm not sure why it happens too. So far, I've seen it only once, and the only prominent thing which can be seen from logs is that there were _a lot_ of disconnected events in a row

> One question, are you wrapping the client with some socket library like DataDog or something to that effect that would audit the data from the socket?

No, we don't. Is there anything specific you'd want to see in the logs? That can be added on our side, of course.

> since you seem to be able to reproduce this, could I post a fix that you could test for me?

I couldn't reproduce it locally, and I saw it on a production server only once.

> One more - what authentication is configured on this cluster, is this a NATS resolver?

It's running via [nats-operator + single secret](https://github.com/nats-io/nats-operator#using-a-single-secret-with-the-explicit-configuration). I don't know what it uses under the hood.


------

The fix looks good to me. Since I can't reproduce it locally, I'll try to run it on servers for some time to see what we will get

**Comment by aricart** [comment]
> @aricart
> 
> > are these servers direct connected or are they behind some other load balancer or something to that effect?
> 
> They are behind the balancer, so there's only 1 server URL specified in the config

This is why the error is spurious - something is happening between NATS server and kubernetes. If the client is run with the `debug` option (which will emit a lot of chatter, the error that the node socket is emitting will print with `!!!` prefix. 

Need to write a small custom server  to make sure the socket errors surfacing correctly.

**Comment by ArmorDarks** [comment]
> This is why the error is spurious - something is happening between NATS server and kubernetes. If the client is run with the debug option (which will emit a lot of chatter, the error that the node socket is emitting will print with !!! prefix.

I was able to see it only on prod servers and wasn't able to reproduce it locally. I can't place very chatty debug mode on prod servers. Maybe there's a way in NATS lib to enable debug only for specific modules, or only emit errors/warnings?

**Comment by aricart** [comment]
we don't have that mechanism, but obviously I could add something for your testing.

**Comment by ArmorDarks** [comment]
It's not directly related, but it seems NATS lib could you a more granular logging system so that it would be possible to enable/disable only specific levels and only for a specific module.

That can be done with [debug](https://github.com/debug-js/debug#readme), or the way it's [done in Mediasoup](https://mediasoup.org/documentation/v3/mediasoup/api/#WorkerLogTag), for instance, with levels + tags

**Comment by ArmorDarks** [comment]
@aricart Do you still plan to add necessary logging for you? If yes, is there any ETA? We plan to restart NATS nodes within a few days, which seems to be the only way to trigger those issues on some servers, so I hope we can throw in logging before that.

**Comment by aricart** [comment]
@ArmorDarks I want to put in the changes without the logging, but I could do a version on nats@next that prints any errors we get. Note that the current changes now tell you what closed the client. There was an issue on the property that the framework was looking for the error (node was storing it in the wrong place)

**Comment by aricart** [comment]
do your services check for the nc.closed() promise?

**Comment by aricart** [comment]
Fixed by #526 and https://github.com/nats-io/nats.deno/pull/390

The handling of the socket close, was attempting to flush which then created the crasher you have, this now won't happen.

**Comment by aricart** [comment]
@ArmorDarks `npm install nats@next` should get you `nats@2.8.1-2` 

Changes are here https://github.com/nats-io/nats.js/pull/527

**Comment by aricart** [comment]
@aricart I am going to close this particular issue (optimistic) if we see some other error from the logging, please open a new issue with the new information so we can continue the search. I think the extra steps will prevent the spurrious erros which crashed the client, but still interested in seeing what sort of socket errors you are getting from node.

**Comment by ArmorDarks** [comment]
👍 

> do your services check for the nc.closed() promise? (they should)

@aricart Yes, it looks like this

```ts
      connection.closed().then((error) => {
        if (error) {
          // error logging
        }
        
        // normal logging
      })
```

Should I also catch on it?

**Comment by aricart** [comment]
@ArmorDarks wondering if you have seen it again.

**Comment by ArmorDarks** [comment]
@aricart Thank you for keeping an eye on it. Edge version was deployed yesterday. So far didn't see it. Let's see how it goes through the next week.

**Comment by ArmorDarks** [comment]
@aricart I think we're good with those changes. Today we decommissioned some NATS nodes (which usually triggered those unhandled rejections), but there were zero unexpected non-catchable errors today.

I have seen a few other interesting errors, though (including https://github.com/nats-io/nats.js/issues/531), but I will post them as separate issues.
